### PR TITLE
Add CMake buildfiles

### PR DESCRIPTION
**Pull request by multiplemonomials**
In order for the Amber CMake build system to work, these buildfiles will have to get added.  Note that there is a duplicate of the cpptraj source list in `src/CMakeLists.txt`.  You will have to keep this up to date if `ambpdbfiles` and `cpptrajfiles` change.  I suppose that if you hated the idea of two source lists, it would be possible to parse those source list files with CMake, but it would end up rather ugly.  Hopefully this'll work out fine.  

Now, about readline.  Since other AmberTools programs (gleap) use readline, and the cpptraj-included readline doesn't work on Windows out of the box, I started with a clean slate when I did my CMake conversion.  I found a readline that had been ported to Windows (`JuliaLang/readline` on GitHub), and wrote a full CMake build system with a configuration script for it.  I added it as a top-level Amber Tool (in the `AmberTools/src` directory), and set up the cpptraj CMake scripts to use it if the system readline is not available.  So that's why `cpptraj/src/readline` does not have a CMake build system.

### CODE REVIEW

**Comment by swails** [comment]
run jenkins

**Comment by swails** [comment]
For what it's worth, gleap is abandoned and should never be built (it's not even bundled with a release). 

As far as I know, only cpptraj uses readline (excluding ParmEd which uses the Python bindings).

**Comment by multiplemonomials** [comment]
Well, nobody told me that about gleap until long after I'd made a build system for it.  Eh, it doesn't matter now.

That's kind of why I mentioned that about readline -- I'd like to move my Windows-compatible readline into cpptraj's source tree.  It's an easy change in the CMake build system, but the problem is that cpptraj's build system would have to change to accommodate it.  Currently, you use a static config.h for readline generated by some configure script from the distant past, and just build readline without any sort of platform checks.  I've always felt like this is kind of rickety, and what you'd have to do with the new readline is run the Autotools (or CMake) configure script before building it.

**Comment by hainm** [comment]
> Well, nobody told me that about gleap until long after I'd made a build system for it. 

Just ask. Feel free to do that if you have any questions. AMBER developers focus most of their time in doing research rather doing proper software development, so there is no detail wiki about things should/shouldn't do.

**Comment by multiplemonomials** [comment]
Hai, you raised a good point that it's kind of misleading to have CMake files in each submodule that don't actually work to build the program.  After sleeping on it, I realized that there are some relatively simple changes that I can make to the main Amber CMake configuration scripts that would allow them to be re-used in different configurations with each submodule.  A lot of what these scripts do is handle MPI, OpenMP, CUDA, and the various Amber dependencies (zlib, arpack, and such), and this functionality is needed by cpptraj and the other C++/C/Fortran submodules.  Basically, I could make the existing cpptraj CMake script into one that could either be run in standalone mode (with most of the same features and options as the Amber build system), or run as part of the larger build system.

@drroe would this be useful to you?  I should be able to replicate the functionality of the current configure script and makefiles without too much trouble using what I've already written for Amber.  I would add the CMake scripts as a submodule, so that I don't have to maintain multiple different versions of the scripts.  The CMake scripts module would have about 50 files and is 300KB.

@hainm I mentioned earlier that the Amber build system as a whole cannot build with Microsoft Visual C++, and I think we've established that MinGW cannot build pytraj in a way that allows it to link with Python.  *However*, if we give cpptraj a standalone CMake build system, then cpptraj only *will* be able to build with MSVC, and then we can build pytraj with MSVC and produce a Miniconda-compatible Windows build of pytraj.  The only slightly hard part is getting a MSVC build of FFTW3 and Readline, but I can do that with my CMake-converted build systems.  If a Windows build of pytraj is your goal, then this is how to do it.

**Comment by hainm** [comment]
> ain Amber CMake configuration scripts that would allow them to be re-used in different configurations with each submodule.

Yes, this should be a way to go.

> I think we've established that MinGW cannot build pytraj in a way that allows it to link with Python

Note that if the Python is built for MinGW and you use that Python to build pytraj, it's fine. We hare working version here: https://github.com/Amber-MD/pytraj/blob/master/appveyor.yml (shown before). The difference here is that you can not use MinGW-built pytraj with Miniconda Python compiled via MVC.

>The only slightly hard part is getting a MSVC build of FFTW3 and Readline, but I can do that with my CMake-converted build systems.

pytraj do not use libreadline, so we can ignore that via `-noreadline`. And we can ignore fftw3 too.
(cpptraj and libcpptraj should be compiled differently).

**Comment by multiplemonomials** [comment]
> And we can ignore fftw3 too.

Actually, I think we do need fftw3.  If it doesn't have fftw3, cpptraj falls back to PubFFT, which is a Fortran program and requires a Fortran compiler, which is not available easily on Windows.

**Comment by drroe** [comment]
Sorry for the long delay in replying. I've been out sick for a few days. First off, thanks for all the work on this!

>  cpptraj-included readline doesn't work on Windows out of the box

I'm all for updating the readline library that is included with cpptraj to a windows-compatible one, but the one you listed (https://github.com/JuliaLang/readline) is GPL V2 which I believe is not compatible with V3 unless the words "or any later version" are included (see [here](https://en.wikipedia.org/wiki/GNU_General_Public_License#Compatibility_and_multi-licensing)). I could be wrong about that though - @swails can correct me if I'm wrong, he has a better grasp of license interactions I think.

> Currently, you use a static config.h for readline generated by some configure script from the distant past,

A few words in my defense here :-). As it was originally distributed, cpptraj was entirely part of AmberTools, which itself uses a "configure script from the distance past". This has vexed generations of Amber programmers (including @swails and @hainm I believe), but there is a reason behind this madness. One of the goals of Amber has been to get it to build on as many UNIX-like systems as possible, and those systems don't always have things like cmake, or believe it or not good/working versions of autotools. I.e. the only way to have a reliable build system was to make it yourself. This would cause problems with e.g. NetCDF _because_ of autotools - in particular I remember some issues on Mercury (teragrid) and the IBM Bluegene that was at BNL. Because I wanted cpptraj to build wherever AmberTools needed to build I followed a similar strategy.

> @drroe would this be useful to you? I should be able to replicate the functionality of the current configure script and makefiles without too much trouble using what I've already written for Amber.

Absolutely, and thank you for taking this on! This is a very welcome addition. I will take a few moments to look over what has been done and then we can go about merging.

**Comment by drroe** [comment]
> Note that there is a duplicate of the cpptraj source list in src/CMakeLists.txt. You will have to keep this up to date if ambpdbfiles and cpptrajfiles change. I suppose that if you hated the idea of two source lists, it would be possible to parse those source list files with CMake, but it would end up rather ugly. 

@multiplemonomials Why not have a simple script that parses `ambpdbfiles` and `cpptrajfiles` into a file with CMake-compatible format? I think Cmakelists supports something like an `include` directive right? The CMake versions of the files can then be included in `src/CMakeLists.txt`. Does that make sense?

**Comment by multiplemonomials** [comment]
Actually, I just realized how to do it.  All I have to do is a couple of string replaces (`string(REPLACE "AMBPDBFILES=" set(AMBPDBFILES "`) on `ambpdbfiles` and `cpptrajfiles`, and then I'll be able to parse them as CMake with an include directive.  The CMake build system can automatically convert the files when it's run.

Also, I'm sorry, I think I was unclear about the "configure script from the distant past".  I wasn't talking about your configure script for cpptraj.  I meant that a long time in the past, someone ran readline's Autotools configure script, and saved the result as `cpptraj/src/readline/config.h`.  This is now used whenever the internal readline is built, and causes some portability problems.  Actually, what I could do is take my readline config.h generator that I wrote for the other readline's CMake build system, and drop it into cpptraj's readline, and it would probably fix a lot of the issues.

Also, do you have the authority to create repositories on the AmberMD Github organization?  Could you create a CMakeConfigScripts repository for me?

Oh, I got the cpptraj standalone CMake build system working last night.  It has the same interface as the Amber build system.  It can also create Mac, Windows, and Linux packages for cpptraj. Once the repository for the CMake scripts is created, I can commit my changes to my cpptraj fork and you can try it out.

Hope you're ferling better.

**Comment by hainm** [comment]
@multiplemonomials https://github.com/Amber-MD/CMakeConfigScripts

**Comment by multiplemonomials** [comment]
@hainm Thanks for creating it.  I accepted your invitation to join the organization, but I cannot seem to push to the new repository.  Can you add me to the collaborators list?

**Comment by hainm** [comment]
can you try again?

You are in "developers" list but I forgot to enable "developers" to have "write" permission.

**Comment by multiplemonomials** [comment]
Awesome!  It worked!

File: src/Action_NAstruct.cpp [hunk 1]
```diff
@@ -1,1832 +1,1833 @@
...
-          // Synchronize all step data sets from rank.
-          NA_Sync( entry->second.shift_, rank_offsets, rank_frames, rank, in );
-          NA_Sync( entry->second.slide_, rank_offsets, rank_frames, rank, in );
-          NA_Sync( entry->second.rise_, rank_offsets, rank_frames, rank, in );
-          NA_Sync( entry->second.tilt_, rank_offsets, rank_frames, rank, in );
-          NA_Sync( entry->second.roll_, rank_offsets, rank_frames, rank, in );
-          NA_Sync( entry->second.twist_, rank_offsets, rank_frames, rank, in );
-          NA_Sync( entry->second.xdisp_, rank_offsets, rank_frames, rank, in );
-          NA_Sync( entry->second.ydisp_, rank_offsets, rank_frames, rank, in );
-
```
Comment (hainm) [inline src/Action_NAstruct.cpp]: what are the differences?

**Comment by multiplemonomials** [comment]
I'm still working towards the goal of getting a working pytraj Windows build.  Using my submodule build system, I've now managed to build cpptraj for Windows using Visual Studio.  It only required a few minor changes to the source code.  I also built this Windows installer (using fftw and mkl linked statically):
[cpptraj.exe.zip](https://github.com/Amber-MD/cpptraj/files/1175506/cpptraj.exe.zip)

**Comment by multiplemonomials** [comment]
oh, wait what?  It looks like something decided to change all of the line endings.  I'll fix it later.  The changes are replacing `__WIN32` with `_WIN32` , and adding `#include <algorithm>` at the top of about 25 files that use `std::max` or `std::min`

**Comment by hainm** [comment]
I guess you need to change your editor to Unix mode. Or using dos2unix.

**Comment by multiplemonomials** [comment]
...what?  I think something strange happened with the Jenkins build.

```
FATAL: command execution failed
java.io.EOFException
	at java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2638)
	at java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:3113)
	at java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:853)
	at java.io.ObjectInputStream.<init>(ObjectInputStream.java:349)
	at hudson.remoting.ObjectInputStreamEx.<init>(ObjectInputStreamEx.java:48)
	at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:34)
	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:59)
Caused: java.io.IOException: Unexpected termination of the channel
	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:73)
Caused: java.io.IOException: Backing channel 'Mac-Mini' is disconnected.
	at hudson.remoting.RemoteInvocationHandler.channelOrFail(RemoteInvocationHandler.java:192)
	at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:257)
	at com.sun.proxy.$Proxy95.isAlive(Unknown Source)
	at hudson.Launcher$RemoteLauncher$ProcImpl.isAlive(Launcher.java:1051)
	at hudson.Launcher$RemoteLauncher$ProcImpl.join(Launcher.java:1043)
	at hudson.tasks.CommandInterpreter.join(CommandInterpreter.java:155)
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:109)
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:66)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:730)
	at hudson.model.Build$BuildExecution.build(Build.java:206)
	at hudson.model.Build$BuildExecution.doRun(Build.java:163)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:490)
	at hudson.model.Run.execute(Run.java:1735)
	at hudson.matrix.MatrixRun.run(MatrixRun.java:146)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:415)
Build step 'Execute shell' marked build as failure
```

**Comment by drroe** [comment]
run jenkins

**Comment by drroe** [comment]
Darn, just noticed all the dos line endings...going to have to fix that :-(

**Comment by hainm** [comment]
Just revert this

**Comment by drroe** [comment]
@hainm The revert just creates a new PR to cancel out the old one right? I can just make a new PR that will fix the issues, unless there is some other reason to do revert?

**Comment by hainm** [comment]
>  I can just make a new PR that will fix the issues

ah yeah.

**Comment by multiplemonomials** [comment]
Oh sorry about the line endings.  I now realize that back when I was working with Unity, I installed a Visual Studio plugin to convert all line endings to CRLF, and it was still installed when I worked with cpptraj.  Also somehow my Git autocrlf was not set properly.  Sorry about that.

**Comment by drroe** [comment]
Not a problem, was easy to fix.

**Pull request merged**